ui/pages/confirmations/components/confirm/info/approve/approve.tsx

```diff
@@ -1,223 +1,89 @@
-          ? `${t('unlimited')} ${tokenSymbol}`
-          : `${formattedTokenNum} ${tokenSymbol}`
-      }
-      onEditClick={
-        transactionMeta.type === TransactionType.tokenMethodIncreaseAllowance
-          ? () => console.log('TODO on a following ticket')
-          : undefined
-      }
-      editIconClassName="edit-spending-cap"
-    />
-  );
-
-  if (!value) {
-    return null;
-  }
-
-  return (
-    <>
-      <ConfirmInfoRowDivider />
-
-      <ConfirmInfoRow
-        label={t('spendingCap')}
-        tooltip={t('spendingCapTooltipDesc')}
-      >
-        {tokenAmount === UNLIMITED_MSG ? (
-          <Tooltip title={formattedTokenNum}>{SpendingCapElement}</Tooltip>
-        ) : (
-          SpendingCapElement
-        )}
-      </ConfirmInfoRow>
-    </>
-  );
-};
+  const [isOpenEditSpendingCapModal, setIsOpenEditSpendingCapModal] =
+    useState(false);
+  const [customSpendingCap, setCustomSpendingCap] = useState('');
 
-const SpendingCap = () => {
-  const t = useI18nContext();
-
-  const transactionMeta = useSelector(
-    currentConfirmationSelector,
-  ) as TransactionMeta;
-
-  const { userBalance, tokenSymbol, decimals } = useAssetDetails(
-    transactionMeta.txParams.to,
-    transactionMeta.txParams.from,
-    transactionMeta.txParams.data,
+  const { decimals } = useAssetDetails(
+    transactionMeta?.txParams?.to,
+    transactionMeta?.txParams?.from,
+    transactionMeta?.txParams?.data,
   );
 
-  const accountBalance = new BigNumber(userBalance || '0')
-    .dividedBy(new BigNumber(10).pow(Number(decimals || '0')))
-    .toNumber();
-
-  const { pending } = useApproveTokenSimulation(
+  const customTxParamsData = useMemo(() => {
+    return getCustomTxParamsData(transactionMeta?.txParams?.data, {
+      customPermissionAmount: customSpendingCap || '0',
+      decimals,
+    });
+  }, [customSpendingCap, transactionMeta?.txParams?.data, decimals]);
+
+  const { value: estimatedGasLimit } = useAsyncResult(async () => {
+    return await estimateGas({
+      from: transactionMeta.txParams.from,
+      to: transactionMeta.txParams.to,
+      value: transactionMeta.txParams.value,
+      data: customTxParamsData,
+    });
+  }, [customTxParamsData, customSpendingCap, decimals]);
+
+  const [shouldUpdateConfirmation, setShouldUpdateConfirmation] =
+    useState(false);
+  const updateConfirmation = useCallback(() => {
+    if (shouldUpdateConfirmation && estimatedGasLimit) {
+      transactionMeta.txParams.data = customTxParamsData;
+      transactionMeta.txParams.gas = hexToDecimal(estimatedGasLimit as string);
+
+      dispatch(updateCurrentConfirmation(transactionMeta));
+      setShouldUpdateConfirmation(false);
+    }
+  }, [
+    shouldUpdateConfirmation,
+    estimatedGasLimit,
+    customTxParamsData,
     transactionMeta,
-    decimals || '0',
-  );
+    dispatch,
+  ]);
 
-  if (pending) {
-    return <Container isLoading />;
-  }
+  useEffect(() => {
+    updateConfirmation();
```

Comment:
Not sure why we have a callback that is only called from an effect? Would it be more readable to just include the logic directly within the effect?